contrib/scalar/scalar.c

```diff
@@ -19,55 +19,98 @@ static int is_unattended(void) {
 	return git_env_bool("Scalar_UNATTENDED", 0);
 }
 
+/**
+ * Remove the deepest subdirectory in the provided path string. Path must not
+ * include a trailing path separator.
+ */
+static void strbuf_parentdir(struct strbuf *buf)
+{
+	size_t offset = offset_1st_component(buf->buf);
+	char *path_sep = find_last_dir_sep(buf->buf + offset);
+	strbuf_setlen(buf, path_sep ? path_sep - buf->buf : offset);
+}
+
+/**
+ * Given an absolute path at or below the enlistment root, find the
+ * enlistment root and working directory
+ */
+static int find_enlistment(struct strbuf *path,
+			   struct strbuf *enlistment_root)
+{
+	char *base_dir;
+	size_t offset = offset_1st_component(path->buf);
+	do {
+		const size_t len = path->len;
+
+		/* check if currently in enlistment root with src/ workdir */
+		strbuf_addstr(path, "/src/.git");
+		if (is_git_directory(path->buf)) {
+			strbuf_strip_suffix(path, "/.git");
+
+			if (enlistment_root) {
+				strbuf_addbuf(enlistment_root, path);
+				strbuf_parentdir(enlistment_root);
+			}
+
+			return 0;
+		}
+
+		/* reset to original path */
+		strbuf_setlen(path, len);
+
+		/* check if currently in workdir */
+		strbuf_addstr(path, "/.git");
+		if (is_git_directory(path->buf)) {
+			strbuf_setlen(path, len);
+
+			if (enlistment_root) {
+				strbuf_addbuf(enlistment_root, path);
+
+				base_dir = find_last_dir_sep(enlistment_root->buf + offset);
+				if (base_dir && !strcmp(base_dir + 1, "src"))
+					strbuf_setlen(enlistment_root,
+						base_dir - enlistment_root->buf);
+			}
```

Comment:
I _just_ found a nice function called `strip_path_suffix()` that might make this much cheaper:

```c
char *root = strip_path_suffix(path, "src");

/*
 * If the worktree's directory's name is `src`, the enlistment is the
 * parent directory, otherwise it is identical to the worktree.
 */
strbuf_addstr(enlistment_root, root ? root : path);
free(root);
```

Reply:
That is certainly much cleaner - updated!